scripts/runChainTests.ts

```diff
@@ -0,0 +1,195 @@
+import { IProcOpts, StatusCode } from './types';
+
+// Stores all the processes
+const procs: { [key: string]: ChildProcessWithoutNullStreams } = {};
+
+// Set the env variable for SAS_SUBSTRATE_WS_URL
+const setWsUrl = (url: string): void => {
+	process.env.SAS_SUBSTRATE_WS_URL = url;
+};
+
+/**
+ * Launch any given process. It accepts an options object.
+ *
+ * {
+ *   proc => the name of the process to be saved in our cache
+ *   resolver => string: If the stdout contains the resolver it will resolve the process
+ *   resolverStartupErr => string: If the stderr contains the resolver it will resolve the process
+ *   args => an array of args to be attached to the `yarn` command.
+ * }
+ *
+ * @param IProcOpts
+ */
+const launchProcess = async ({
+	proc,
+	resolver,
+	resolverStartupErr,
+	args,
+}: IProcOpts): Promise<StatusCode> => {
+	return new Promise<StatusCode>((resolve, reject) => {
+		const { Success, Failed } = StatusCode;
+
+		const command = 'yarn';
+
+		procs[proc] = spawn(command, args, { detached: true });
+
+		procs[proc].stdout.on('data', (data: Buffer) => {
+			console.log(data.toString());
+
+			if (data.toString().includes(resolver)) {
+				resolve(Success);
+			}
+		});
+
+		procs[proc].stderr.on('data', (data: Buffer) => {
+			console.error(data.toString());
+
+			if (resolverStartupErr && data.toString().includes(resolverStartupErr)) {
+				resolve(Failed);
+			}
+		});
+
+		procs[proc].on('close', () => {
+			resolve(Success);
+		});
+
+		procs[proc].on('error', (err) => {
+			console.log(err);
+			reject(Failed);
+		});
+	});
+};
+
+/**
+ * Launches Sidecar, and if successful it will launch the jest runner. This operation
+ * handles killing all the processes after the jest runner is done.
+ *
+ * @param chain The chain in which to target the e2e tests too.
+ */
+const launchChainTest = async (chain: string): Promise<boolean> => {
+	const { wsUrl, SasStartOpts, JestProcOpts } = config[chain];
+	const { Success } = StatusCode;
+
+	// Set the ws url env var
+	setWsUrl(wsUrl);
+
+	console.log('Launching Sidecar...');
+	const sidecarStart = await launchProcess(SasStartOpts);
+
+	if (sidecarStart === Success) {
+		// Sidecar successfully launched, and jest will now get called
+		console.log('Launching jest...');
+		const jest = await launchProcess(JestProcOpts);
+
+		if (jest === Success) {
+			killAll();
+			return true;
+		} else {
+			killAll();
+			return false;
+		}
+	} else {
+		console.error('Error launching sidecar... exiting...');
+		killAll();
+		process.exit(1);
+	}
+};
+
+// Kill all processes spawned and tracked by this file.
+const killAll = () => {
+	console.log('Killing all processes...');
+	for (const key of Object.keys(procs)) {
+		if (!procs[key].killed) {
+			try {
+				console.log(`Killing ${key}`);
+				// Kill child and all its descendants.
+				process.kill(-procs[key].pid, 'SIGTERM');
+				process.kill(-procs[key].pid, 'SIGKILL');
+			} catch (e) {
+				/**
```

Comment:
could console.error the message, but up to you

Reply:
Yea i thought about it too, but it will throw error every time when using `-procs[key].pid` because there will be some 'pid' that sidecar already closed implicitly but is still within the group. The error is harmless, but the major thing is just making sure all process's close.